solr/core/src/java/org/apache/solr/util/circuitbreaker/CircuitBreakerManager.java

```diff
@@ -135,15 +125,61 @@ public static String toErrorMessage(List<CircuitBreaker> circuitBreakerList) {
    *
    * Any default circuit breakers should be registered here.
    */
+  @Deprecated
   public static CircuitBreakerManager build(PluginInfo pluginInfo) {
+    return build(pluginInfo, null);
+  }
+
+  /**
+   * TODO
+   */
+  public static CircuitBreakerManager build(PluginInfo pluginInfo, SolrResourceLoader solrResourceLoader) {
     boolean enabled = pluginInfo == null ? false : Boolean.parseBoolean(pluginInfo.attributes.getOrDefault("enabled", "false"));
     CircuitBreakerManager circuitBreakerManager = new CircuitBreakerManager(enabled);
 
-    circuitBreakerManager.init(pluginInfo);
+    circuitBreakerManager.init(pluginInfo, solrResourceLoader);
 
     return circuitBreakerManager;
   }
 
+  /**
+   * Initialize with circuit breakers defined in the configuration
+   */
+  public void init(PluginInfo pluginInfo, SolrResourceLoader solrResourceLoader) {
```

Comment:
Did you consider `implements ResourceLoaderAware PluginInfoInitialized` annotations instead of hardwired init in CC?

Reply:
Yes, I did. The main challenge was that both (ResourceLoader and PluginInfo list) need to be present when CircuitBreakerManager can be initialised. If ResourceLoaderAware and PluginInfoInitialized are used, then CircuitBreakerManager will have to maintain state for both until they are both present, and then initialise itself. This seems a bit of an overkill?

Reply:
I can see that. But perhaps worth giving it a try? The state machine could be quite simple. Each `inform()` would set an instance var and call `init()`, and `init()` would just exit if it lacks either or the two?

Reply:
@janhoy I am not sure if I understand the nuances of ResourceLoaderAware correctly, but have pushed an iteration. Please see and let me know if that is what you meant.

Reply:
I have also given `ResourceLoaderAware` use a try but couldn't quite figure out how to make it work and if it's actually needed in this scenario? The https://github.com/apache/lucene-solr/blob/releases/lucene-solr/8.9.0/solr/core/src/java/org/apache/solr/core/SolrResourceLoader.java#L823-L849 caused asserts to trigger if `SolrCoreAware` or `ResourceLoaderAware` is used and the comments guide against additions to the list (though admittedly I haven't yet looked at the SOLR-8311 referenced there).

`PluginInfoInitialized` use alone seems to be sufficient if we're happy to have the circuit breakers register with the circuit breaker manager similar to how the metric producers register with the metric manager?

Reply:
Agreed